File: dlp/cpp/prac4_RDP.cpp
#include <iostream>
#include <string>

class Parser {
public:
    bool parseE(std::string& input) {
        if (input.empty()) {
            return false;
        }
        if (input[0] == 'a') {
            input.erase(0, 1);  // Consume 'a'
            return parseX(input);
        }
        return false;
    }

    bool parseX(std::string& input) {
        if (input.empty()) {
            return false;
        }
        if (input[0] == 'c') {
            input.erase(0, 1);  // Consume 'c'
            return true;
        } else if (input[0] == '+') {
            input.erase(0, 1);  // Consume '+'
            if (input[0] == 'a') {
                input.erase(0, 1);  // Consume 'a'
                return parseX(input);
            }
        }
        return false;
    }
};

int main() {
    Parser parser;
    std::string input;

    std::cout << "Enter an input string: ";
    std::cin >> input;

    if (parser.parseE(input) && input.empty()) {
        std::cout << "Valid string" << std::endl;
    } else {
        std::cout << "Invalid string" << std::endl;
    }

    return 0;
}

File: dlp/python_dlp/prac4_RDP.py
class Parser:
    def __init__(self, input_string):
        self.input = input_string

    def parseE(self):
        if self.input and self.input[0] == 'a':
            self.input = self.input[1:]  # Consume 'a'
            return self.parseX()
        return False

    def parseX(self):
        if not self.input:
            return False
        if self.input[0] == 'c':
            self.input = self.input[1:]  # Consume 'c'
            return True
        elif self.input[0] == '+':
            self.input = self.input[1:]  # Consume '+'
            if self.input and self.input[0] == 'a':
                self.input = self.input[1:]  # Consume 'a'
                return self.parseX()
        return False

def main():
    input_string = input("Enter an input string: ")
    parser = Parser(input_string)

    if parser.parseE() and not parser.input:
        print("Valid string")
    else:
        print("Invalid string")

if __name__ == "__main__":
    main()
